### ISSUE TITLE
Override not working as anticipated

### ISSUE THREAD

**Issue by Sparticuz** [eslint/eslint#13585]
**eslint 7.6.0**

I'm trying to add TS & JS linting, but with different rules, settings, etc... While making the right config and looking at the docs, I was a little confused, so I asked in discord about it. First, the documentation for overrides was a little sparse, so I'm not sure what the intended functionality is. I would think that the overrides would override any config 'higher up' the config, but it seems that this is not really the case. It seems like the entire rules block is overridden.

Example: In my config, I have `"no-restricted-syntax"` in the base config, so I can use for-of (which I think the original settings for that are in airbnb-base, and they disallow for-of). This rule works fine in JS files, but when I split out the overrides for TS files, I would expect the rule to also apply to them, but it doesn't (verified by --print-config on a TS file).

#### Basically, my question/change request would be: Does the Overrides block override the rules block, or does/should it be augmenting it? Do the documents needs to be updated to make the correct implementation more clear? Or is this a bug?

<details>
  <summary>eslint config</summary>
  
  ```javascript
    module.exports = {
  env: {
    es6: true,
    es2017: true,
    es2020: true,
    node: true,
  },
  extends: [
    "airbnb-base",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:promise/recommended",
    "plugin:node/recommended",
    "plugin:unicorn/recommended",
    "prettier",
    "plugin:prettier/recommended",
  ],
  // Ignore my dist folders so I don't lint transpiled files
  ignorePatterns: "dist/**/*",
  overrides: [
    {
      // Test Javascript Files Override Block
      files: ["**/tests/*.js"],
      rules: {
        "node/no-unpublished-require": 0,
      },
    },
    {
      // Typescript Override Block
      extends: [
        "airbnb-base",
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
        "plugin:import/errors",
        "plugin:import/warnings",
        "plugin:import/typescript",
        "plugin:promise/recommended",
        "plugin:node/recommended",
        "plugin:unicorn/recommended",
        "prettier",
        "plugin:prettier/recommended",
        "prettier/@typescript-eslint",
      ],
      files: ["**/*.ts", "**/*.tsx"],
      parser: "@typescript-eslint/parser",
      parserOptions: {
        project: "tsconfig.json",
      },
      plugins: [
        "@typescript-eslint",
        "typescript-sort-keys",
        "import",
        "node",
        "prettier",
        "promise",
        "unicorn",
      ],
      rules: {
        // Make sure we don't require .ts and .tsx
        "import/extensions": [
          "error",
          "ignorePackages",
          {
            ts: "never",
            tsx: "never",
          },
        ],
        // Typescript uses Imports for transpiling
        "node/no-unsupported-features/es-syntax": [
          "error",
          {
            ignores: ["modules"],
          },
        ],
      },
      settings: {
        "import/parsers": {
          "@typescript-eslint/parser": [".ts", ".tsx"],
        },
        "import/resolver": {
          // user <root>/tsconfig.json
          "typescript": {
            "alwaysTryTypes": true,
          }
        },
        node: {
          // Make sure we are looking for Typescript files as well
          tryExtensions: [".js", ".json", ".node", ".ts", ".d.ts"],
        },
      },
    },
  ],
  parser: "esprima",
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: "module",
  },
  plugins: ["import", "node", "prettier", "promise", "unicorn"],
  rules: {
    // disallow certain syntax forms, allows for-of
    // http://eslint.org/docs/rules/no-restricted-syntax
    "no-restricted-syntax": [
      "error",
      "ForInStatement",
      "LabeledStatement",
      "WithStatement",
    ],
    "sort-keys": [
      "error",
      "asc",
      {
        natural: true,
      },
    ],
    // Allow aws-sdk to exist in optionalDependencies and be used in production
    "import/no-extraneous-dependencies": [
      "error",
      {
        optionalDependencies: true,
      },
    ],
  },
};
  ```
</details>

<details>
  <summary>relevant sections of a --print-config on a ts file</summary>
  
  ```javascript
...
"no-restricted-syntax": [
      "error",
      {
        "selector": "ForInStatement",
        "message": "for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array."
      },
      {
        "selector": "ForOfStatement",
        "message": "iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations."
      },
      {
        "selector": "LabeledStatement",
        "message": "Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand."
      },
      {
        "selector": "WithStatement",
        "message": "`with` is disallowed in strict mode because it makes code impossible to predict and optimize."
      }
    ],
...
"sort-keys": [
      "off",
      "asc",
      {
        "caseSensitive": false,
        "natural": true
      }
    ],
...
  ```
</details

**Comment by kaicataldo** [comment]
>  Does the Overrides block override the rules block, or does/should it be augmenting it? Do the documents needs to be updated to make the correct implementation more clear? Or is this a bug?

The overrides should be merged into the top-level config. If that's not happening, it sounds like something might be misconfigured or we there's been some kind of regression in the config system. Any chance you could make a Git repository with the minimal reproduction case so we can get a better idea of what's going?

Re documentation, PRs are welcome to improve it!

**Comment by kaicataldo** [comment]
On an unrelated note, I'd recommend using the default parser rather than `esprima`. You can remove the `parser` line in the config and it'll use our default parser :)

**Comment by kaicataldo** [comment]
I'm unable to reproduce this behavior, so a reproduction case would be really helpful here.

**Comment by Sparticuz** [comment]
> On an unrelated note, I'd recommend using the default parser rather than `esprima`. You can remove the `parser` line in the config and it'll use our default parser :)

I thought that was the default parser 😆 

> I'm unable to reproduce this behavior, so a reproduction case would be really helpful here.

I'm also having trouble reproducing it when I am doing a minimal reproduction. Still working on it though

**Comment by Sparticuz** [comment]
@kaicataldo I believe I've found what's wrong. Could it be a bug or unintended consequence, maybe? In my top level block, I'm extending from `airbnb-base`, then overriding the `"no-restricted-syntax"` rule. Then in my Typescript override block, I'm again, extending `airbnb-base`, however, I'm not overriding the rule that time. I assumed that the rules block would merge with the top level, then apply itself last over the top of the extends blocks, but it seems like eslint takes the top level extends, applies the top level rules, then takes any override blocks, runs the extends over top of the top level (with included rules), then overrides the rules of that with the override rules.

In a nutshell, 
```
top level extends -> top level rules      -> override extends -> override rules
(airbnb-base)     -> no-restricted-syntax -> (airbnb-base)    -> no rule override
```

So that seems to by why the rule isn't being enforced, because it's being overridden a second time because of my override extends.

- Does `extends` replace the top level `extends` or merge with it like `rules` does? (Order of the `extends` would matter)

- Would altering the way this works have alot of consequences? ie `override extends -> top level rules -> override rules`

https://github.com/Sparticuz/eslint-overrides

**Comment by kaicataldo** [comment]
I'm a bit confused - what benefit does extending the original shared config in the overrides provide? That would override any configuration in your top-level `rules` configuration. It sounds like the configuration for `no-restricted-syntax` set in `airbnb-base` is overriding what you have defined in your top level `rules` config.

If you only want `*.js` files to be linted with your `no-restricted-syntax` configuration, could you just set that in `overrides`?

**Comment by Sparticuz** [comment]
Currently, I have the top level config, which would apply to all js files, then in the `overrides`, I have it set to be just ts files. I suppose I could probably have nothing in the top level config, then just have multiple `overrides` for each file type, but I would lose inheritance that way. (not that I have it now...) I do it this way because otherwise I have to set a bunch of eslint ignore rules on the top of each javascript file for all the typescript rules that don't apply to js.

I want everything to use `no-restricted-syntax`, but when I set it in top level, then use `extends airbnb-base` again, the rule gets overridden.

Thinking about it some more, because of the way eslint applies overrides, I don't really even need to 're-extend' most of the configs, I really just need to add any typescript related extends. Then override any rules that I would need to there.

**Comment by kaicataldo** [comment]
You should never have to "re-extend' something, since it's already being applied at the top level :)

**Comment by Sparticuz** [comment]
I'm seeing that now. I was assuming that if I used `extends` in an `override`, it would overwrite the top level config instead of ... extending it :D

I think we're good now.
